src/language-server/project/internal.ts

```diff
@@ -0,0 +1,346 @@
+    const { includes = [], excludes = [] } = config.client;
+
+    this.documentsByFile = new Map();
+
+    this.fileSet.pushIncludes(includes);
+    // We do not want to include the local schema file in our list of documents
+    this.fileSet.pushExcludes([
+      ...excludes,
+      ...this.getRelativeLocalSchemaFilePaths(),
+    ]);
+
+    this.schemaProvider = schemaProviderFromConfig(config, clientIdentity);
+    const { engine } = config;
+    if (engine.apiKey) {
+      this.engineClient = new ApolloEngineClient(
+        engine.apiKey!,
+        engine.endpoint,
+        clientIdentity,
+      );
+    }
+  }
+
+  public resolveSchema(config: SchemaResolveConfig): Promise<GraphQLSchema> {
+    this.lastLoadDate = +new Date();
+    return this.schemaProvider.resolveSchema(config);
+  }
+
+  public resolveFederatedServiceSDL() {
+    return this.schemaProvider.resolveFederatedServiceSDL();
+  }
+
+  public onSchemaChange(handler: NotificationHandler<GraphQLSchema>) {
+    this.lastLoadDate = +new Date();
+    return this.schemaProvider.onSchemaChange(handler);
+  }
+
+  includesFile(uri: DocumentUri) {
+    return this.fileSet.includesFile(uri);
+  }
+
+  allIncludedFiles() {
+    return this.fileSet.allFiles();
+  }
+
+  async scanAllIncludedFiles() {
+    await this.loadingHandler.handle(
+      `Loading queries for ${this.displayName}`,
+      (async () => {
+        for (const filePath of this.allIncludedFiles()) {
+          const uri = URI.file(filePath).toString();
+
+          // If we already have query documents for this file, that means it was either
+          // opened or changed before we got a chance to read it.
+          if (this.documentsByFile.has(uri)) continue;
+
+          this.fileDidChange(uri);
+        }
+      })(),
+    );
+  }
+
+  fileDidChange(uri: DocumentUri) {
+    const filePath = URI.parse(uri).fsPath;
+    const extension = extname(filePath);
+    const languageId = fileAssociations[extension];
+
+    // Don't process files of an unsupported filetype
+    if (!languageId) return;
+
+    // Don't process directories. Directories might be named like files so
+    // we have to explicitly check.
+    if (!lstatSync(filePath).isFile()) return;
+
+    const contents = readFileSync(filePath, "utf8");
+    const document = TextDocument.create(uri, languageId, -1, contents);
+    this.documentDidChange(document);
+  }
+
+  fileWasDeleted(uri: DocumentUri) {
+    this.removeGraphQLDocumentsFor(uri);
+    this.checkForDuplicateOperations();
+  }
+
+  documentDidChange = (document: TextDocument) => {
+    const documents = extractGraphQLDocuments(
+      document,
+      this.config.client && this.config.client.tagName,
+    );
+    if (documents) {
+      this.documentsByFile.set(document.uri, documents);
+      this.invalidate();
+    } else {
+      this.removeGraphQLDocumentsFor(document.uri);
+    }
+    this.checkForDuplicateOperations();
+  };
+
+  checkForDuplicateOperations = throttle(
```

Comment:
This is now throttled. Before, events leading up to `documentDidChange` were debounced, but that lead to file updates being completely lost if more than one document changed.